core/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java

```diff
@@ -241,6 +246,58 @@ static MetaData updateMetaDataWithRoutingTable(MetaData oldMetaData, RoutingTabl
     }
 
     /**
+     * Possibly removes delay markers from shards and reroutes if markers have been deleted.
+     * <p>
+     * If the same instance of the routing table is returned, then no change has been made.</p>
+     */
+    public RoutingAllocation.Result applyDelayedShards(ClusterState clusterState) {
+        return applyDelayedShards(clusterState, true);
+    }
+
+    public RoutingAllocation.Result applyDelayedShards(ClusterState clusterState, boolean withReroute) {
+        RoutingNodes routingNodes = getMutableRoutingNodes(clusterState);
+        RoutingAllocation allocation = new RoutingAllocation(allocationDeciders, routingNodes, clusterState.nodes(), clusterInfoService.getClusterInfo(), currentNanoTime());
+        boolean changed = removeDelayMarkers(allocation);
+        if (changed == false) {
+            return new RoutingAllocation.Result(false, clusterState.routingTable(), clusterState.metaData());
+        }
+        if (withReroute) {
+            reroute(allocation);
+        }
+        RoutingAllocation.Result result = buildChangedResult(allocation);
+        logClusterHealthStateChange(
+            new ClusterStateHealth(clusterState),
+            new ClusterStateHealth(clusterState.getMetaData(), result.routingTable()),
+            "assign delayed unassigned shards"
+        );
+        return result;
+
+    }
+
+    /**
+     * Removes delay markers from unassigned shards based on current time stamp. Returns true if markers were removed.
+     */
+    public static boolean removeDelayMarkers(RoutingAllocation allocation) {
```

Comment:
I expect to be tested as it's static, but I don't see test? I think its good to have one. Also, it seems it can be package private.